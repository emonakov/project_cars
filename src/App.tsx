import React, { lazy, Suspense } from 'react';
import { ThemeProvider } from 'styled-components';
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';

import ContextProvider from './components/Context';
import Header from './components/Header';
import Footer from './components/Footer';
import Fallback from './components/Fallback';
import theme from './config/theme';

const Home = lazy(() => import('./components/Page/Home'));
const CarDetails = lazy(() => import('./components/Page/CarDetails'));

const App: React.FC = () => (
    <ThemeProvider theme={theme}>
        <Router>
            <Suspense fallback={<Fallback />}>
                <Header />
                <ContextProvider>
                    <Switch>
                        <Route path="/" exact component={Home} />
                        <Route path="/car/:stockId(\d+)" exact component={CarDetails} />
                    </Switch>
                </ContextProvider>
                <Footer />
            </Suspense>
        </Router>
    </ThemeProvider>
);

export default App;
