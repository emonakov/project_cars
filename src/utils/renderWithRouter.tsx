import React from 'react';
import { Router } from 'react-router-dom';
import { render } from '@testing-library/react';
import { createMemoryHistory } from 'history';

const history = createMemoryHistory();

const renderWithRouter = (Component: JSX.Element, initialRoute = '/') => {
    history.push(initialRoute);

    return render(
        <Router history={history}>
            {Component}
        </Router>,
    );
};

export default renderWithRouter;
