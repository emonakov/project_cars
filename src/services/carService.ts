import axios from 'redaxios';

import { buildApiUrl } from '../utils/api';

export const fetchCar = async (stockId: string): Promise<any> => {
    const { data } = await axios.get(buildApiUrl(`api/cars/${stockId}`));
    const { car } = JSON.parse(data);

    return car;
};
