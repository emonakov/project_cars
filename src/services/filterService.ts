import axios from 'redaxios';

import { buildApiUrl } from '../utils/api';

export const fetchFilters = async (): Promise<any> => {
    const manufacturersUrl = buildApiUrl('api/manufacturers');
    const colorsUrl = buildApiUrl('api/colors');

    const [manResponse, colResponse] = await Promise.all([axios.get(manufacturersUrl), axios.get(colorsUrl)]);
    const [{ manufacturers }, { colors }] = [JSON.parse(manResponse.data), JSON.parse(colResponse.data)];

    return { manufacturers, colors };
};

export const fetchCarsWithFilters = async (options: any[]): Promise<any> => {
    const carsUrl = options
        .filter(([, paramValue]) => Boolean(paramValue))
        .reduce((acc, [paramKey, paramValue]) => `${acc}&${paramKey}=${paramValue}`, 'api/cars?');

    const { data } = await axios.get(buildApiUrl(carsUrl));

    return JSON.parse(data);
};
