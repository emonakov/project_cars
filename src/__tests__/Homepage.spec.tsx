import React from 'react';
import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import App from '../App';
import { fetchFilters, fetchCarsWithFilters } from '../services/filterService';
import renderWithRouter from '../utils/renderWithRouter';
import { manufacturers } from './__mock__/manufacturers.mock.json';
import { colors } from './__mock__/colors.mock.json';
import cars from './__mock__/cars.mock.json';

jest.mock('../services/filterService');

describe('Homepage', () => {
    beforeEach(() => {
        (fetchFilters as jest.Mock).mockImplementation(() => Promise.resolve({ manufacturers, colors }));

        (fetchCarsWithFilters as jest.Mock).mockImplementation(() => Promise.resolve(cars.Audi));
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('renders the initial page', () => {
        renderWithRouter(<App />);
        expect(screen.getByText(/car sales/i)).toBeInTheDocument();
    });

    it('renders homepage and waits for all the elements', async () => {
        renderWithRouter(<App />);

        await screen.findByTestId('manufacturers');
        expect(screen.getByTestId('manufacturers')).toBeInTheDocument();
        expect(screen.getByTestId('colors')).toBeInTheDocument();
    });

    it('renders the page and changes the filter values', async () => {
        renderWithRouter(<App />);
        (fetchCarsWithFilters as jest.Mock).mockImplementation(() => Promise.resolve(cars.BMW));

        await screen.findByTestId('manufacturers');
        userEvent.type(screen.getByTestId('manufacturers'), 'BMW');
        userEvent.type(screen.getByTestId('colors'), 'white');
        userEvent.click(screen.getByTestId('filter'));

        await screen.findByText(/66926/);
        expect(screen.getByText(/66926/)).toBeInTheDocument();
        expect(screen.getByText(/65365/)).toBeInTheDocument();
        expect(screen.getByText(/65365/)).toMatchSnapshot();
        expect(screen.getByText(/66926/)).toMatchSnapshot();
    });

    it('does not render car list if it is empty', async () => {
        (fetchCarsWithFilters as jest.Mock).mockImplementation(() => Promise.resolve({}));

        renderWithRouter(<App />);

        await screen.findByTestId('car-list');
        expect(screen.getByTestId('car-list')).toBeEmpty();
    });
});
