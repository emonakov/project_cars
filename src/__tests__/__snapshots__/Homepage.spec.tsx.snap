// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Homepage renders the page and changes the filter values 1`] = `
<span
  class="MuiTypography-root MuiTypography-caption MuiTypography-gutterBottom MuiTypography-displayBlock"
>
  Stock # 65365
  <span
    class="sc-fzplWN beffhE"
  >
     - 186,866 km
  </span>
  <span
    class="sc-fznyAO looylO"
  >
     - Petrol - white
  </span>
  <div
    class="MuiCardActions-root MuiCardActions-spacing"
  >
    <a
      class="sc-AxirZ iLjBBz sc-fznKkj bkajBi"
      data-testid="car-65365"
      href="/car/65365"
    >
      View details
    </a>
  </div>
</span>
`;

exports[`Homepage renders the page and changes the filter values 2`] = `
<span
  class="MuiTypography-root MuiTypography-caption MuiTypography-gutterBottom MuiTypography-displayBlock"
>
  Stock # 66926
  <span
    class="sc-fzplWN beffhE"
  >
     - 188,218 km
  </span>
  <span
    class="sc-fznyAO looylO"
  >
     - Diesel - white
  </span>
  <div
    class="MuiCardActions-root MuiCardActions-spacing"
  >
    <a
      class="sc-AxirZ iLjBBz sc-fznKkj bkajBi"
      data-testid="car-66926"
      href="/car/66926"
    >
      View details
    </a>
  </div>
</span>
`;
